internal/storage/pgx_driver.go

```diff
@@ -0,0 +1,246 @@
+
+	var pgErr *pgconn.PgError
+	err = d.createTables()
+	if errors.As(err, &pgErr) && pgerrcode.IsConnectionException(pgErr.Code) {
+		logger.Log.Warn(pgerrcode.ConnectionFailure, zap.Int("attemp", 1))
+		return err
+	} else if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *pgxDriver) Close() error {
+	d.conn.Close()
+	return nil
+}
+
+func (d *pgxDriver) Ping() error {
+	return d.conn.Ping(context.Background())
+}
+
+func (d *pgxDriver) Save() error {
+	return nil
+}
+
+func (d *pgxDriver) Restore() error {
+	return nil
+}
+
+func (d *pgxDriver) Update(mtype, mname, mval string) error {
+	var sqlScript string
+	switch mtype {
+	case CounterType:
+		sqlScript = `
+		INSERT INTO metrics (mtype, mname, mvalue) 
+		VALUES ($1, $2, $3) 
+		ON CONFLICT (mname) DO 
+		UPDATE SET mvalue = EXCLUDED.mvalue::int + metrics.mvalue::int
+		`
+	case GaugeType:
+		sqlScript = `
+		INSERT INTO metrics (mtype, mname, mvalue) 
+		VALUES ($1, $2, $3) 
+		ON CONFLICT (mname) DO 
+		UPDATE SET mvalue = EXCLUDED.mvalue
+		`
+	}
+	_, err := d.exec(context.Background(), sqlScript, mtype, mname, mval)
+	return err
+}
+
+func (d *pgxDriver) Get(mtype, mname string) (string, error) {
+	if mtype == "" || mname == "" {
+		return "", errors.New("invalid metric type")
+	}
+	row := d.queryRow(context.Background(), `
+	SELECT mvalue FROM metrics WHERE mtype=$1 AND mname=$2
+	`, mtype, mname)
+	var mval sql.NullString
+	err := row.Scan(&mval)
+	if err != nil {
+		return "", err
+	}
+	if mval.Valid {
+		return mval.String, nil
+	}
+	return "", errors.New("not found")
+}
+
+// TODO: нужны тесты, не хватает времени
+func (d *pgxDriver) GetAll() Data {
+	var data Data
+	rows, err := d.queryRows(context.Background(), `SELECT mtype, mname, mvalue FROM metrics`)
+	if err != nil {
+		logger.Log.Error(err.Error())
+		return data
+	}
+	defer rows.Close()
+
+	counters := make(map[string]int64)
+	gauges := make(map[string]float64)
+	for rows.Next() {
+		var mtype, mname, mvalue string
+		if err = rows.Scan(&mtype, &mname, &mvalue); err != nil {
+			logger.Log.Error(err.Error())
+			return data
+		}
+		switch mtype {
+		case CounterType:
+			var value int64
+			if value, err = strconv.ParseInt(mvalue, 10, 64); err != nil {
+				return data
+			}
+			counters[mname] = value
+		case GaugeType:
+			var value float64
+			if value, err = strconv.ParseFloat(mvalue, 64); err != nil {
+				return data
+			}
+			gauges[mname] = value
+		}
+	}
+	err = rows.Err()
+	if err != nil {
+		return data
+	}
+	data.Counters = counters
+	data.Gauges = gauges
+	return data
+}
+
+func (d *pgxDriver) UpdateAll(data Data) error {
+	ctx := context.Background()
+	counters := counters2Any(data.Counters)
+	gauges := gauges2Any(data.Gauges)
+	if err := d.updateAny(ctx, counters, CounterType); err != nil {
+		return err
+	}
+	if err := d.updateAny(ctx, gauges, GaugeType); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *pgxDriver) updateAny(ctx context.Context, m AnyMetrics, mtype string) error {
```

Comment:
Не хочу занудствовать по поводу английского языка, но any буквально означает любой, что прямо противоположно по смыслу тому, что делает код.

```suggestion
func (d *pgxDriver) updateAll(ctx context.Context, m AnyMetrics, mtype string) error {
```

Reply:
Возможно, но эта функция вызывается в функции UpdateAll(), если переименовать то теряется понимание что делает функция, а при просмотре всего файла а не конкретного кусочка все становиться понятно